apps/payments/transferwise.py

```diff
@@ -1,7 +1,121 @@
+    def inner(f):
+        webhook_handlers[type] = f
+        return f
+
+    return inner
+
+
+@csrf.exempt
+@payments.route("/transferwise-webhook", methods=["POST"])
+def transferwise_webhook():
+    valid_signature = verify_signature(request.data, request.headers["X-Signature"],)
+    if not valid_signature:
+        logger.exception("Error verifying TransferWise webhook signature")
+        abort(400)
+
+    event_type = request.json.get("event_type")
+    try:
+        try:
+            handler = webhook_handlers[event_type]
+        except KeyError as e:
+            handler = webhook_handlers[None]
+        return handler(event_type, request.json)
+    except Exception as e:
+        logger.exception("Unhandled exception during TransferWise webhook")
+        logger.info("Webhook data: %s", request.data)
+        abort(500)
+
+
+@webhook("balances#credit")
+def transferwise_balance_credit(event_type, event):
+    profile_id = event.get("data", {}).get("resource", {}).get("profile_id")
+    if profile_id is None:
+        logger.exception("Missing profile_id in TransferWise webhook")
+        logger.info("Webhook data: %s", request.data)
+        abort(400)
+
+    borderless_account_id = event.get("data", {}).get("resource", {}).get("id")
+    if borderless_account_id is None:
+        logger.exception("Missing borderless_account_id in TransferWise webhook")
+        logger.info("Webhook data: %s", request.data)
+        abort(400)
+
+    currency = event.get("data", {}).get("currency")
+    if currency is None:
+        logger.exception("Missing currency in TransferWise webhook")
+        logger.info("Webhook data: %s", request.data)
+        abort(400)
+
+    # Find the TransferWise bank account in the application database
+    bank_account = BankAccount.query.filter_by(
+        borderless_account_id=borderless_account_id, active=True
+    ).first()
+    if not bank_account:
+        logger.exception("Could not find bank account for borderless_account_id")
+        logger.info("Webhook data: %s", request.data)
+        abort(400)
+
+    # Retrieve an account transaction statement for the past week
+    client = pytransferwise.Client()
+    interval_end = datetime.now()
+    interval_start = interval_end - timedelta(days=7)
+    statement = client.borderless_accounts.statement(
+        profile_id=profile_id,
+        account_id=borderless_account_id,
+        currency=currency,
+        interval_start=interval_start.isoformat() + "Z",
+        interval_end=interval_end.isoformat() + "Z",
+    )
+
+    # Retrieve or construct transactions for each credit in the statement
+    txns = []
+    for transaction in statement.transactions:
+        if transaction.type != "CREDIT":
+            continue
+
+        # Attempt to find transaction in the application database
+        txn = BankTransaction.query.filter(
+            account_id=bank_account.id,
+            posted=transaction.date,
+            type=transaction.details.type.lower(),
```

Comment:
From the TransferWise [borderless account statement](https://api-docs.transferwise.com/#borderless-accounts-get-account-statement) docs:

```
transactions[n].details.type
---
CARD, CONVERSION, DEPOSIT, TRANSFER, MONEY_ADDED
```

This will have an impact on [transaction type validation](https://github.com/emfcamp/Website/blob/7e839497fcdc3cc27c27c936009e07008cc41004/apps/base/tasks_banking.py#L140) during reconciliation (assuming code is re-used).